### PR TITLE
Reactive context menus take two

### PR DESCRIPTION
**Pull request by SomeHats**
In #2697, we make context menu overrides reactive in response to a user request. They wanted to add some extra context menu items depending on the state of the editor, but couldn't get those to update reactively.

Because we could now re-run overrides many times, that caused an issue where as we were mutating the schema, we would add many new 'cursor chat' entries instead of just the one, so we reverted that part of the original PR in #2775.

Ideally our overrides wouldn't mutate, but they do, so whatever. We can fix that issue by making sure the overrides always get a fresh copy of the original schema, but there's another more subtle issue: because our cursor chat override depends on whether the editor is in course pointer mode via `getInstanceState`, it updates every time the cursor moves creating a lot of spurious react re-renders.

The quick-fix solution introduced here is to just cache this value in a `computed`. It does still render more often that it should though - for example, because `getSelectedShapes` depends on `currentPageState` which includes hover info, it re-runs whenever the hovered shape changes.

Instead of the caching here, we could introduce a `UserPreferenceManager` style helper for instance & page state. These two records both contain largely unrelated data, some of which updates very frequently. If all those property accesses were cached, we'd avoid that re-rendering issue.

### Change Type

- [x] `patch` — Bug fix

### Test Plan

- [x]  Unit Tests
- Manual testing to make sure we're not re-rendering too often

### CODE REVIEW

**Review by ds300 — COMMENTED** [review]
> Instead of the caching here, we could introduce a UserPreferenceManager style helper for instance & page state

I'm pretty fine with this.

The main question is what to do about setters. We currently have history-integrated 'commands' for updating those records, so we could do one of the following

1. Have per-field setters on the helper class that use the history command under the hood, and maybe expose the history-related flags (`TLCommandHistoryOptions`)
2. Don't have any setters and just expect people to call the existing method on the root class.

I'd lean towards 1. I think.

File: packages/utils/src/lib/function.ts [hunk 1]
```diff
@@ -59,3 +59,17 @@ export function omitFromStackTrace<Args extends Array<unknown>, Return>(
...
+ */
+export function weakCache<T extends object, U>(fn: (arg: T) => U): (arg: T) => U {
+	const cache = new WeakMap<T, U>()
+	return (arg: T) => {
+		if (!cache.has(arg)) {
+			cache.set(arg, fn(arg))
+		}
+		return cache.get(arg)!
+	}
+}
```
[the remark this answers is not in the document]
Comment (ds300) [reply packages/utils/src/lib/function.ts:75]: looking at this closer, i think we can delete WeakMapCache. We don't use any of it's fancy methods, and I'm 99.9% sure it's a no-op in the the one place where we're using so we can just rip it out entirely.

**Comment by steveruizok** [comment]
Should we close this now that composable menus have landed? or move the discussion somewhere else? Glad that the issue finally led to using React for our menus.

**Comment by SomeHats** [comment]
yeah this can be closed now!

**Pull request closed without merging**